File: 백준/Gold/3665. 최종 순위/최종 순위.py
# 이번년도 순위 높아졌으면 앞에 들어온다. 
# 잘못된 정보가 있을수도..? 모순 발생할듯
# 올해 순위를 완벽히 못 만들수도 있다.

from collections import deque

test_case = int(input())

for i in range(test_case):
    n = int(input())
    num_list = list(map(int, input().split(' ')))
    adj = [[] for _ in range(n+1)]
    indegree = [0] * (n + 1)

    for i in range(n):
        for j in range(i + 1, n):
            a = num_list[i]
            b = num_list[j]
            adj[a].append(b)
            indegree[b] += 1


    m = int(input())

    for j in range(m):
        num1, num2 = map(int, input().split(' '))
    
        # 방향 바꾸기
        if num2 in adj[num1]:
            adj[num1].remove(num2)
            adj[num2].append(num1)
            indegree[num2] -= 1
            indegree[num1] += 1
        else:
            adj[num2].remove(num1)
            adj[num1].append(num2)
            indegree[num1] -= 1
            indegree[num2] += 1
            
    q = deque()

    for i in range(1, n+1):
        if indegree[i] == 0:
            q.append(i)
    
    result = []
    certain = True #확실한 순위
    cycle = False # 사이클 여부

    for _ in range(n):
        if not q:
            cycle = True
            break
        if len(q) > 1:
            certain = False

        now = q.popleft()
        result.append(now)

        for nxt in adj[now]:
            indegree[nxt] -= 1
            if indegree[nxt] == 0:
                q.append(nxt)

    if cycle:
        print("IMPOSSIBLE")
    elif not certain:
        print("?")
    else:

        print(' '.join(map(str, result)))

File: 백준/Gold/3665. 최종 순위/최종 순위.cc
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

int main() {
    int T;
    cin >> T;

    while (T--) {
        int n;
        cin >> n;

        vector<int> num_list(n + 1);
        vector<vector<int>> adj(n + 1);
        vector<int> indegree(n + 1, 0);

        for (int i = 1; i <= n; i++) {
            cin >> num_list[i];
        }

        // 간선 생성
        for (int i = 1; i <= n; i++) {
            for (int j = i + 1; j <= n; j++) {
                int a = num_list[i];
                int b = num_list[j];
                adj[a].push_back(b);
                indegree[b]++;
            }
        }

        int m;
        cin >> m;

        // 상대 순위 바뀐 쌍 반영
        for (int i = 0; i < m; i++) {
            int a, b;
            cin >> a >> b;

            // 원래 a -> b 였던 경우
            auto it = find(adj[a].begin(), adj[a].end(), b);
            if (it != adj[a].end()) {
                adj[a].erase(it);
                adj[b].push_back(a);
                indegree[b]--;
                indegree[a]++;
            } else {
                // 원래 b -> a 였던 경우
                auto it2 = find(adj[b].begin(), adj[b].end(), a);
                if (it2 != adj[b].end()) {
                    adj[b].erase(it2);
                    adj[a].push_back(b);
                    indegree[a]--;
                    indegree[b]++;
                }
            }
        }

        queue<int> q;
        for (int i = 1; i <= n; i++) {
            if (indegree[i] == 0) {
                q.push(i);
            }
        }

        vector<int> result;
        bool certain = true;
        bool cycle = false;

        for (int i = 0; i < n; i++) {
            if (q.empty()) {
                cycle = true;
                break;
            }
            if (q.size() > 1) {
                certain = false;
            }

            int now = q.front();
            q.pop();
            result.push_back(now);

            for (int nxt : adj[now]) {
                indegree[nxt]--;
                if (indegree[nxt] == 0) {
                    q.push(nxt);
                }
            }
        }

        if (cycle) {
            cout << "IMPOSSIBLE\n";
        } else if (!certain) {
            cout << "?\n";
        } else {
            for (int i = 0; i < result.size(); i++) {
                cout << result[i];
                if (i != result.size() - 1) cout << " ";
            }
            cout << "\n";
        }
    }

    return 0;
}
